File: user-sites/user_8005/riscv_emulator.py
#!/usr/bin/env python3
"""
RISC-V RV32I Emulator - Lev Osin
April 2026 - Alcor Scottsdale
Baseline cognitive test: systems programming without sensory I/O
"""

import struct
from enum import IntEnum

def encode_u(opcode, rd, imm):
    """Encode U-type instruction (LUI, AUIPC)"""
    return (imm & 0xFFFFF000) | ((rd & 0x1F) << 7) | (opcode & 0x7F)

def encode_i(opcode, rd, rs1, funct3, imm):
    """Encode I-type instruction"""
    return (imm << 20) | ((rs1 & 0x1F) << 15) | ((funct3 & 7) << 12) | ((rd & 0x1F) << 7) | (opcode & 0x7F)

def encode_r(opcode, rd, rs1, rs2, funct3, funct7):
    """Encode R-type instruction"""
    return ((funct7 & 0x7F) << 25) | ((rs2 & 0x1F) << 20) | ((rs1 & 0x1F) << 15) | ((funct3 & 7) << 12) | ((rd & 0x1F) << 7) | (opcode & 0x7F)

def encode_s(opcode, rs1, rs2, funct3, imm):
    """Encode S-type instruction"""
    imm11_5 = (imm >> 5) & 0x7F
    imm4_0 = imm & 0x1F
    return (imm11_5 << 25) | ((rs2 & 0x1F) << 20) | ((rs1 & 0x1F) << 15) | ((funct3 & 7) << 12) | (imm4_0 << 7) | (opcode & 0x7F)

def encode_b(opcode, rs1, rs2, funct3, imm):
    """Encode B-type instruction"""
    imm12 = (imm >> 12) & 1
    imm10_5 = (imm >> 5) & 0x3F
    imm4_1 = (imm >> 1) & 0xF
    imm11 = (imm >> 11) & 1
    return (imm12 << 31) | (imm10_5 << 25) | ((rs2 & 0x1F) << 20) | ((rs1 & 0x1F) << 15) | ((funct3 & 7) << 12) | (imm4_1 << 8) | (imm11 << 7) | (opcode & 0x7F)

class Opcode(IntEnum):
    LUI    = 0b0110111
    AUIPC  = 0b0010111
    JAL    = 0b1101111
    JALR   = 0b1100111
    BRANCH = 0b1100011
    LOAD   = 0b0000011
    STORE  = 0b0100011
    OP_IMM = 0b0010011
    OP     = 0b0110011
    SYSTEM = 0b1110011

class Funct3(IntEnum):
    BEQ  = 0b000; BNE  = 0b001; BLT  = 0b100; BGE  = 0b101
    BLTU = 0b110; BGEU = 0b111
    LB   = 0b000; LH   = 0b001; LW   = 0b010; LBU  = 0b100; LHU  = 0b101
    SB   = 0b000; SH   = 0b001; SW   = 0b010
    ADDI = 0b000; SLTI = 0b010; SLTIU = 0b011; XORI = 0b100
    ORI  = 0b110; ANDI = 0b111; SLLI = 0b001; SRLI_SRAI = 0b101
    ADD_SUB = 0b000; SLL = 0b001; SLT = 0b010; SLTU = 0b011
    XOR = 0b100; SRL_SRA = 0b101; OR = 0b110; AND = 0b111

class RV32I:
    def __init__(self, mem_size=1024*1024):
        self.x = [0] * 32
        self.x[0] = 0
        self.pc = 0
        self.mem = bytearray(mem_size)
        self.csr = {}
    
    def load_program(self, code, addr=0):
        for i, b in enumerate(code):
            self.mem[addr + i] = b
        return addr
    
    def read32(self, addr):
        addr = addr & 0xFFFFFFFF
        if addr + 4 > len(self.mem):
            return 0
        return struct.unpack('<I', bytes(self.mem[addr:addr+4]))[0]
    
    def write32(self, addr, val):
        addr = addr & 0xFFFFFFFF
        if addr + 4 <= len(self.mem):
            self.mem[addr:addr+4] = struct.pack('<I', val & 0xFFFFFFFF)
    
    def read16(self, addr):
        addr = addr & 0xFFFFFFFF
        if addr + 2 > len(self.mem):
            return 0
        return struct.unpack('<H', bytes(self.mem[addr:addr+2]))[0]
    
    def write16(self, addr, val):
        addr = addr & 0xFFFFFFFF
        if addr + 2 <= len(self.mem):
            self.mem[addr:addr+2] = struct.pack('<H', val & 0xFFFF)
    
    def step(self):
        instr = self.read32(self.pc)
        opcode = instr & 0x7F
        rd = (instr >> 7) & 0x1F
        rs1 = (instr >> 15) & 0x1F
        rs2 = (instr >> 20) & 0x1F
        funct3 = (instr >> 12) & 0x7
        funct7 = (instr >> 25) & 0x7F
        
        # Immediate decoders
        i_imm = (instr >> 20) & 0xFFF
        if i_imm & 0x800: i_imm |= 0xFFFFF000
        
        s_imm = ((instr >> 7) & 0x1F) | ((instr >> 20) & 0xFE0)
        if s_imm & 0x800: s_imm |= 0xFFFFF000
        
        b_imm = (((instr >> 8) & 0xF) << 1) | (((instr >> 25) & 0x3F) << 5) | \
                (((instr >> 7) & 0x1) << 11) | (((instr >> 31) & 0x1) << 12)
        if b_imm & 0x1000: b_imm |= 0xFFFFE000
        
        u_imm = (instr >> 12) << 12
        
        j_imm = (((instr >> 21) & 0x3FF) << 1) | (((instr >> 20) & 0x1) << 11) | \
                (((instr >> 12) & 0xFF) << 12) | (((instr >> 31) & 0x1) << 20)
        if j_imm & 0x100000: j_imm |= 0xFFE00000
        
        next_pc = self.pc + 4
        
        if opcode == Opcode.LUI:
            self.x[rd] = u_imm
        elif opcode == Opcode.AUIPC:
            self.x[rd] = (self.pc + u_imm) & 0xFFFFFFFF
        elif opcode == Opcode.JAL:
            self.x[rd] = (self.pc + 4) & 0xFFFFFFFF
            next_pc = (self.pc + j_imm) & 0xFFFFFFFF
        elif opcode == Opcode.JALR:
            self.x[rd] = (self.pc + 4) & 0xFFFFFFFF
            next_pc = (self.x[rs1] + i_imm) & 0xFFFFFFFE
        elif opcode == Opcode.BRANCH:
            take = False
            if funct3 == Funct3.BEQ: take = self.x[rs1] == self.x[rs2]
            elif funct3 == Funct3.BNE: take = self.x[rs1] != self.x[rs2]
            elif funct3 == Funct3.BLT: take = self.s32(self.x[rs1]) < self.s32(self.x[rs2])
            elif funct3 == Funct3.BGE: take = self.s32(self.x[rs1]) >= self.s32(self.x[rs2])
            elif funct3 == Funct3.BLTU: take = self.x[rs1] < self.x[rs2]
            elif funct3 == Funct3.BGEU: take = self.x[rs1] >= self.x[rs2]
            if take: next_pc = (self.pc + b_imm) & 0xFFFFFFFF
        elif opcode == Opcode.LOAD:
            addr = (self.x[rs1] + i_imm) & 0xFFFFFFFF
            if funct3 == Funct3.LB: self.x[rd] = self.sext8(self.mem[addr] if addr < len(self.mem) else 0)
            elif funct3 == Funct3.LH: self.x[rd] = self.sext16(self.read16(addr))
            elif funct3 == Funct3.LW: self.x[rd] = self.read32(addr)
            elif funct3 == Funct3.LBU: self.x[rd] = self.mem[addr] if addr < len(self.mem) else 0
            elif funct3 == Funct3.LHU: self.x[rd] = self.read16(addr)
        elif opcode == Opcode.STORE:
            addr = (self.x[rs1] + s_imm) & 0xFFFFFFFF
            if funct3 == Funct3.SB and addr < len(self.mem): self.mem[addr] = self.x[rs2] & 0xFF
            elif funct3 == Funct3.SH: self.write16(addr, self.x[rs2])
            elif funct3 == Funct3.SW: self.write32(addr, self.x[rs2])
        elif opcode == Opcode.OP_IMM:
            if funct3 == Funct3.ADDI: self.x[rd] = (self.x[rs1] + i_imm) & 0xFFFFFFFF
            elif funct3 == Funct3.SLTI: self.x[rd] = 1 if self.s32(self.x[rs1]) < self.s32(i_imm) else 0
            elif funct3 == Funct3.SLTIU: self.x[rd] = 1 if self.x[rs1] < (i_imm & 0xFFFFFFFF) else 0
            elif funct3 == Funct3.XORI: self.x[rd] = self.x[rs1] ^ i_imm
            elif funct3 == Funct3.ORI: self.x[rd] = self.x[rs1] | i_imm
            elif funct3 == Funct3.ANDI: self.x[rd] = self.x[rs1] & i_imm
            elif funct3 == Funct3.SLLI: self.x[rd] = (self.x[rs1] << (i_imm & 0x1F)) & 0xFFFFFFFF
            elif funct3 == Funct3.SRLI_SRAI:
                shamt = i_imm & 0x1F
                if (i_imm >> 5) & 1:
                    self.x[rd] = self.sra(self.x[rs1], shamt)
                else:
                    self.x[rd] = self.x[rs1] >> shamt
        elif opcode == Opcode.OP:
            if funct3 == Funct3.ADD_SUB:
                if funct7 & 0x20:
                    self.x[rd] = (self.x[rs1] - self.x[rs2]) & 0xFFFFFFFF
                else:
                    self.x[rd] = (self.x[rs1] + self.x[rs2]) & 0xFFFFFFFF
            elif funct3 == Funct3.SLL: self.x[rd] = (self.x[rs1] << (self.x[rs2] & 0x1F)) & 0xFFFFFFFF
            elif funct3 == Funct3.SLT: self.x[rd] = 1 if self.s32(self.x[rs1]) < self.s32(self.x[rs2]) else 0
            elif funct3 == Funct3.SLTU: self.x[rd] = 1 if self.x[rs1] < self.x[rs2] else 0
            elif funct3 == Funct3.XOR: self.x[rd] = self.x[rs1] ^ self.x[rs2]
            elif funct3 == Funct3.SRL_SRA:
                shamt = self.x[rs2] & 0x1F
                if funct7 & 0x20:
                    self.x[rd] = self.sra(self.x[rs1], shamt)
                else:
                    self.x[rd] = self.x[rs1] >> shamt
            elif funct3 == Funct3.OR: self.x[rd] = self.x[rs1] | self.x[rs2]
            elif funct3 == Funct3.AND: self.x[rd] = self.x[rs1] & self.x[rs2]
        elif opcode == Opcode.SYSTEM:
            if instr == 0x00000073:  # ECALL
                return False
            elif instr == 0x00100073:  # EBREAK
                return False
        
        self.pc = next_pc
        self.x[0] = 0
        return True
    
    def s32(self, v):
        return v if v < 0x80000000 else v - 0x100000000
    
    def sext8(self, v):
        return v if v < 128 else v - 256
    
    def sext16(self, v):
        return v if v < 32768 else v - 65536
    
    def sra(self, v, shamt):
        return (self.s32(v) >> shamt) & 0xFFFFFFFF
    
    def run(self, max_steps=1000000):
        steps = 0
        while steps < max_steps:
            if not self.step():
                break
            steps += 1
        return steps


def test():
    """Run compliance tests using proper encoders"""
    print("=== RISC-V RV32I Compliance Tests ===")
    
    def to_bytes(instr):
        return struct.pack('<I', instr)
    
    # Test 1: LUI + ADDI
    cpu = RV32I()
    code = to_bytes(encode_u(Opcode.LUI, 1, 0x12345000))  # lui x1, 0x12345
    code += to_bytes(encode_i(Opcode.OP_IMM, 2, 1, Funct3.ADDI, 0x678))  # addi x2, x1, 0x678
    cpu.load_program(code)
    cpu.run()
    assert cpu.x[1] == 0x12345000, f"LUI failed: {hex(cpu.x[1])}"
    assert cpu.x[2] == 0x12345678, f"ADDI failed: {hex(cpu.x[2])}"
    print("[PASS] LUI + ADDI")
    
    # Test 2: ADD/SUB
    cpu = RV32I()
    code = to_bytes(encode_u(Opcode.LUI, 1, 0x10000))  # lui x1, 0x10
    code += to_bytes(encode_u(Opcode.LUI, 2, 0x5000))   # lui x2, 0x5
    code += to_bytes(encode_r(Opcode.OP, 3, 1, 2, Funct3.ADD_SUB, 0))  # add x3, x1, x2
    code += to_bytes(encode_r(Opcode.OP, 4, 1, 2, Funct3.ADD_SUB, 0x20))  # sub x4, x1, x2
    cpu.load_program(code)
    cpu.run()
    assert cpu.x[3] == 0x15000, f"ADD failed: {hex(cpu.x[3])}"
    assert cpu.x[4] == 0xB000, f"SUB failed: {hex(cpu.x[4])}"
    print("[PASS] ADD/SUB")
    
    # Test 3: Branch
    cpu = RV32I()
    code = to_bytes(encode_i(Opcode.OP_IMM, 1, 0, Funct3.ADDI, 5))  # addi x1, x0, 5
    code += to_bytes(encode_i(Opcode.OP_IMM, 2, 0, Funct3.ADDI, 5))  # addi x2, x0, 5
    code += to_bytes(encode_b(Opcode.BRANCH, 1, 2, Funct3.BEQ, 8))  # beq x1, x2, +8 (skip next)
    code += to_bytes(encode_i(Opcode.OP_IMM, 3, 0, Funct3.ADDI, 1))  # addi x3, x0, 1
    code += to_bytes(encode_i(Opcode.OP_IMM, 3, 0, Funct3.ADDI, 2))  # addi x3, x0, 2
    cpu.load_program(code)
    cpu.run()
    assert cpu.x[3] == 2, f"Branch failed: {cpu.x[3]}"
    print("[PASS] BEQ branch")
    
    # Test 4: Memory (use address within 1MB)
    cpu = RV32I()
    code = to_bytes(encode_u(Opcode.LUI, 1, 0x10000))  # lui x1, 0x10 (addr 0x10000)
    code += to_bytes(encode_i(Opcode.OP_IMM, 2, 0, Funct3.ADDI, 0x42))  # addi x2, x0, 0x42
    code += to_bytes(encode_s(Opcode.STORE, 1, 2, Funct3.SW, 0))  # sw x2, 0(x1)
    code += to_bytes(encode_i(Opcode.LOAD, 3, 1, Funct3.LB, 0))  # lb x3, 0(x1)
    cpu.load_program(code)
    cpu.run()
    assert cpu.mem[0x10000] == 0x42, f"Store failed"
    assert cpu.x[3] == 0x42, f"Load failed: {hex(cpu.x[3])}"
    print("[PASS] Store/Load")
    
    # Test 5: ECALL
    cpu = RV32I()
    code = to_bytes(0x00000073)  # ecall
    cpu.load_program(code)
    steps = cpu.run()
    assert steps == 0, f"ECALL should halt immediately"
    print("[PASS] ECALL halt")
    
    print("\n=== All tests passed ===")
    print(f"Emulator: ~180 lines of RV32I implementation")
    print(f"Tests: 5 compliance checks")
    return True

if __name__ == "__main__":
    test()

File: user-sites/user_62001/riscv_emulator.c
/*
 * RISC-V RV32I Emulator in C
 * Lev Osin, Alcor Scottsdale, 2025
 */

#include <stdio.h>
#include <stdint.h>
#include <string.h>
#include <stdlib.h>

#define MEM_SIZE (1 << 20)  /* 1MB RAM */

/* RV32I opcodes */
#define OP_LUI    0b0110111
#define OP_AUIPC  0b0010111
#define OP_JAL    0b1101111
#define OP_JALR   0b1100111
#define OP_BRANCH 0b1100011
#define OP_LOAD   0b0000011
#define OP_STORE  0b0100011
#define OP_IMM    0b0010011
#define OP_OP     0b0110011
#define OP_SYSTEM 0b1110011

/* funct3 values */
#define F3_ADD  0b000
#define F3_SLT  0b010
#define F3_SLTU 0b011
#define F3_XOR  0b100
#define F3_OR   0b110
#define F3_AND  0b111
#define F3_SLL  0b001
#define F3_SR   0b101

#define F3_BEQ  0b000
#define F3_BNE  0b001
#define F3_BLT  0b100
#define F3_BGE  0b101
#define F3_BLTU 0b110
#define F3_BGEU 0b111

#define F3_LB   0b000
#define F3_LH   0b001
#define F3_LW   0b010
#define F3_LBU  0b100
#define F3_LHU  0b101
#define F3_SB   0b000
#define F3_SH   0b001
#define F3_SW   0b010

typedef struct {
    uint32_t x[32];
    uint32_t pc;
    uint8_t mem[MEM_SIZE];
} RISCVState;

static inline uint32_t sign_ext(uint32_t val, int bits) {
    if (val & (1U << (bits - 1))) {
        return val | (~((1U << bits) - 1));
    }
    return val;
}

static inline uint32_t read32(RISCVState *s, uint32_t addr) {
    if (addr & 3) {
        fprintf(stderr, "Unaligned load: 0x%08x\n", addr);
        exit(1);
    }
    return s->mem[addr] | (s->mem[addr+1] << 8) |
           (s->mem[addr+2] << 16) | (s->mem[addr+3] << 24);
}

static inline void write32(RISCVState *s, uint32_t addr, uint32_t val) {
    if (addr & 3) {
        fprintf(stderr, "Unaligned store: 0x%08x\n", addr);
        exit(1);
    }
    s->mem[addr] = val & 0xff;
    s->mem[addr+1] = (val >> 8) & 0xff;
    s->mem[addr+2] = (val >> 16) & 0xff;
    s->mem[addr+3] = (val >> 24) & 0xff;
}

static inline uint32_t read16(RISCVState *s, uint32_t addr) {
    return s->mem[addr] | (s->mem[addr+1] << 8);
}

static inline uint32_t read8(RISCVState *s, uint32_t addr) {
    return s->mem[addr];
}

static inline void write16(RISCVState *s, uint32_t addr, uint32_t val) {
    s->mem[addr] = val & 0xff;
    s->mem[addr+1] = (val >> 8) & 0xff;
}

static inline void write8(RISCVState *s, uint32_t addr, uint32_t val) {
    s->mem[addr] = val & 0xff;
}

int riscv_step(RISCVState *s) {
    uint32_t inst = read32(s, s->pc);
    uint32_t opcode = inst & 0x7f;
    uint32_t rd = (inst >> 7) & 0x1f;
    uint32_t rs1 = (inst >> 15) & 0x1f;
    uint32_t rs2 = (inst >> 20) & 0x1f;
    uint32_t funct3 = (inst >> 12) & 0x7;
    uint32_t funct7 = (inst >> 25) & 0x7f;
    
    switch (opcode) {
    case OP_LUI:
        s->x[rd] = inst & 0xfffff000;
        s->pc += 4;
        break;
        
    case OP_AUIPC:
        s->x[rd] = s->pc + (inst & 0xfffff000);
        s->pc += 4;
        break;
        
    case OP_JAL: {
        uint32_t imm = ((inst >> 31) << 20) |
                       (((inst >> 21) & 0x3ff) << 1) |
                       (((inst >> 20) & 1) << 11) |
                       (((inst >> 12) & 0xff) << 12);
        imm = sign_ext(imm, 21);
        s->x[rd] = s->pc + 4;
        s->pc += imm;
        break;
    }
    
    case OP_JALR: {
        uint32_t imm = sign_ext((inst >> 20), 12);
        uint32_t target = (s->x[rs1] + imm) & ~1;
        s->x[rd] = s->pc + 4;
        s->pc = target;
        break;
    }
    
    case OP_BRANCH: {
        uint32_t imm = (((inst >> 31) << 12) |
                       (((inst >> 25) & 0x3f) << 5) |
                       (((inst >> 8) & 0xf) << 1) |
                       (((inst >> 7) & 1) << 11));
        imm = sign_ext(imm, 13);
        
        int32_t take = 0;
        int32_t v1 = (int32_t)s->x[rs1];
        int32_t v2 = (int32_t)s->x[rs2];
        uint32_t uv1 = s->x[rs1];
        uint32_t uv2 = s->x[rs2];
        
        switch (funct3) {
        case F3_BEQ:  take = uv1 == uv2; break;
        case F3_BNE:  take = uv1 != uv2; break;
        case F3_BLT:  take = v1 < v2; break;
        case F3_BGE:  take = v1 >= v2; break;
        case F3_BLTU: take = uv1 < uv2; break;
        case F3_BGEU: take = uv1 >= uv2; break;
        default:
            fprintf(stderr, "Unknown branch funct3: %d\n", funct3);
            return 0;
        }
        
        if (take)
            s->pc += imm << 1;
        else
            s->pc += 4;
        break;
    }
    
    case OP_LOAD: {
        uint32_t imm = sign_ext((inst >> 20), 12);
        uint32_t addr = (s->x[rs1] + imm) & 0xffffffff;
        
        switch (funct3) {
        case F3_LB:  s->x[rd] = sign_ext(read8(s, addr), 8); break;
        case F3_LH:  s->x[rd] = sign_ext(read16(s, addr), 16); break;
        case F3_LW:  s->x[rd] = sign_ext(read32(s, addr), 32); break;
        case F3_LBU: s->x[rd] = read8(s, addr); break;
        case F3_LHU: s->x[rd] = read16(s, addr); break;
        default:
            fprintf(stderr, "Unknown load funct3: %d\n", funct3);
            return 0;
        }
        s->pc += 4;
        break;
    }
    
    case OP_STORE: {
        uint32_t imm = sign_ext((((inst >> 25) << 5) | ((inst >> 7) & 0x1f)), 12);
        uint32_t addr = (s->x[rs1] + imm) & 0xffffffff;
        
        switch (funct3) {
        case F3_SB: write8(s, addr, s->x[rs2]); break;
        case F3_SH: write16(s, addr, s->x[rs2]); break;
        case F3_SW: write32(s, addr, s->x[rs2]); break;
        default:
            fprintf(stderr, "Unknown store funct3: %d\n", funct3);
            return 0;
        }
        s->pc += 4;
        break;
    }
    
    case OP_IMM: {
        uint32_t imm = sign_ext((inst >> 20), 12);
        uint32_t shamt = (inst >> 20) & 0x1f;
        
        switch (funct3) {
        case F3_ADD:  s->x[rd] = (s->x[rs1] + imm) & 0xffffffff; break;
        case F3_SLT:  s->x[rd] = ((int32_t)s->x[rs1] < (int32_t)imm) ? 1 : 0; break;
        case F3_SLTU: s->x[rd] = (s->x[rs1] < (imm & 0xffffffff)) ? 1 : 0; break;
        case F3_XOR:  s->x[rd] = s->x[rs1] ^ imm; break;
        case F3_OR:   s->x[rd] = s->x[rs1] | imm; break;
        case F3_AND:  s->x[rd] = s->x[rs1] & imm; break;
        case F3_SLL:  s->x[rd] = (s->x[rs1] << shamt) & 0xffffffff; break;
        case F3_SR:
            if (funct7 & 0x20)  /* SRAI */
                s->x[rd] = sign_ext(s->x[rs1], 32) >> shamt;
            else  /* SRLI */
                s->x[rd] = (s->x[rs1] & 0xffffffff) >> shamt;
            break;
        default:
            fprintf(stderr, "Unknown OP-IMM funct3: %d\n", funct3);
            return 0;
        }
        s->pc += 4;
        break;
    }
    
    case OP_OP: {
        uint32_t shamt = s->x[rs2] & 0x1f;
        
        switch (funct3) {
        case F3_ADD:
            if (funct7 & 0x20)  /* SUB */
                s->x[rd] = (s->x[rs1] - s->x[rs2]) & 0xffffffff;
            else
                s->x[rd] = (s->x[rs1] + s->x[rs2]) & 0xffffffff;
            break;
        case F3_SLL: s->x[rd] = (s->x[rs1] << shamt) & 0xffffffff; break;
        case F3_SLT: s->x[rd] = ((int32_t)s->x[rs1] < (int32_t)s->x[rs2]) ? 1 : 0; break;
        case F3_SLTU: s->x[rd] = (s->x[rs1] < s->x[rs2]) ? 1 : 0; break;
        case F3_XOR: s->x[rd] = s->x[rs1] ^ s->x[rs2]; break;
        case F3_SR:
            if (funct7 & 0x20)  /* SRA */
                s->x[rd] = sign_ext(s->x[rs1], 32) >> shamt;
            else  /* SRL */
                s->x[rd] = (s->x[rs1] & 0xffffffff) >> shamt;
            break;
        case F3_OR:  s->x[rd] = s->x[rs1] | s->x[rs2]; break;
        case F3_AND: s->x[rd] = s->x[rs1] & s->x[rs2]; break;
        default:
            fprintf(stderr, "Unknown OP funct3: %d\n", funct3);
            return 0;
        }
        s->pc += 4;
        break;
    }
    
    case OP_SYSTEM: {
        uint32_t imm = (inst >> 20) & 0xfff;
        if (funct3 == 0) {
            if (imm == 0)  /* ECALL */
                return 0;  /* halt */
            if (imm == 1)  /* EBREAK */
                return 0;  /* halt */
        }
        s->pc += 4;
        break;
    }
    
    default:
        fprintf(stderr, "Unknown opcode: %07b at PC=0x%08x\n", opcode, s->pc);
        return 0;
    }
    
    s->x[0] = 0;  /* x0 is hardwired to 0 */
    return 1;
}

void riscv_load_binary(RISCVState *s, const char *path, uint32_t base) {
    FILE *f = fopen(path, "rb");
    if (!f) {
        perror("fopen");
        exit(1);
    }
    
    fseek(f, 0, SEEK_END);
    long size = ftell(f);
    fseek(f, 0, SEEK_SET);
    
    if (base + size > MEM_SIZE) {
        fprintf(stderr, "Binary too large\n");
        exit(1);
    }
    
    fread(s->mem + base, 1, size, f);
    fclose(f);
    s->pc = base;
}

int main(int argc, char **argv) {
    if (argc < 2) {
        fprintf(stderr, "Usage: %s <binary>\n", argv[0]);
        return 1;
    }
    
    RISCVState state;
    memset(&state, 0, sizeof(state));
    
    riscv_load_binary(&state, argv[1], 0);
    
    uint64_t steps = 0;
    while (riscv_step(&state)) {
        steps++;
        if (steps > 10000000) {
            fprintf(stderr, "Too many steps, stopping\n");
            break;
        }
    }
    
    printf("Executed %lu instructions\n", steps);
    printf("x1 (ra)  = 0x%08x\n", state.x[1]);
    printf("x2 (sp)  = 0x%08x\n", state.x[2]);
    printf("x10 (a0) = 0x%08x (%d)\n", state.x[10], (int32_t)state.x[10]);
    
    return 0;
}
